File: src/2025-02-24-deepclaude-rust/java/exercises/practice/alphametics/src/main/java/Alphametics.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

class Alphametics {

    public static class UnsolvablePuzzleException extends Exception {
        public UnsolvablePuzzleException() {
            super("No solution exists.");
        }
    }

    private final List<String> leftOperands;
    private final String rightOperand;
    private final Set<Character> leadingLetters;
    private final List<Character> uniqueLetters;

    Alphametics(String userInput) {
        // Split input into left and right parts
        String[] parts = userInput.split("=");
        if (parts.length != 2) {
            throw new IllegalArgumentException("Equation must have exactly one '='");
        }

        String[] leftParts = parts[0].split("\\+");
        leftOperands = Arrays.stream(leftParts)
                .map(String::trim)
                .filter(s -> !s.isEmpty())
                .collect(Collectors.toList());
        rightOperand = parts[1].trim();

        if (leftOperands.isEmpty()) {
            throw new IllegalArgumentException("Must have at least one operand on left side");
        }

        // Collect all unique letters and leading letters
        leadingLetters = new HashSet<>();
        Set<Character> allLetters = new HashSet<>();

        for (String operand : leftOperands) {
            if (operand.isEmpty()) {
                throw new IllegalArgumentException("Operand cannot be empty");
            }
            leadingLetters.add(operand.charAt(0));
            for (char c : operand.toCharArray()) {
                allLetters.add(c);
            }
        }

        if (rightOperand.isEmpty()) {
            throw new IllegalArgumentException("Right operand cannot be empty");
        }
        leadingLetters.add(rightOperand.charAt(0));
        for (char c : rightOperand.toCharArray()) {
            allLetters.add(c);
        }

        uniqueLetters = new ArrayList<>(allLetters);
        if (uniqueLetters.size() > 10) {
            throw new IllegalArgumentException("Too many unique letters");
        }
        // Sort unique letters so that leading letters come first
        Collections.sort(uniqueLetters, (a, b) -> {
            boolean aLeading = leadingLetters.contains(a);
            boolean bLeading = leadingLetters.contains(b);
            if (aLeading == bLeading) return Character.compare(a, b);
            return aLeading ? -1 : 1;
        });
    }

    Map<Character, Integer> solve() throws UnsolvablePuzzleException {
        Map<Character, Integer> solution = new HashMap<>();
        boolean[] usedDigits = new boolean[10];
        
        if (backtrack(solution, usedDigits, 0)) {
            return Collections.unmodifiableMap(solution);
        } else {
            throw new UnsolvablePuzzleException();
        }
    }

    private boolean backtrack(Map<Character, Integer> solution, boolean[] usedDigits, int index) {
        if (index == uniqueLetters.size()) {
            return isValid(solution);
        }

        char currentChar = uniqueLetters.get(index);
        for (int digit = 0; digit < 10; digit++) {
            if (usedDigits[digit]) continue;
            if (digit == 0 && leadingLetters.contains(currentChar)) continue;

            usedDigits[digit] = true;
            solution.put(currentChar, digit);
            if (backtrack(solution, usedDigits, index + 1)) {
                return true;
            }
            solution.remove(currentChar);
            usedDigits[digit] = false;
        }
        return false;
    }

    private boolean isValid(Map<Character, Integer> solution) {
        long leftSum = 0;
        for (String operand : leftOperands) {
            leftSum += toNumber(operand, solution);
            if (leftSum < 0) { // Check for overflow
                return false;
            }
        }
        long rightNum = toNumber(rightOperand, solution);
        return leftSum == rightNum;
    }

    private long toNumber(String word, Map<Character, Integer> solution) {
        long num = 0;
        for (char c : word.toCharArray()) {
            num = num * 10 + solution.get(c);
            if (num < 0) { // Check for overflow
                return -1;
            }
        }
        return num;
    }
}

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/alphametics/alphametics.go
package alphametics

import (
	"errors"
	"strings"
)

func Solve(puzzle string) (map[string]int, error) {
	// Parse the puzzle into addends and result
	parts := strings.Split(puzzle, "==")
	if len(parts) != 2 {
		return nil, errors.New("invalid puzzle format")
	}
	leftPart := strings.TrimSpace(parts[0])
	rightPart := strings.TrimSpace(parts[1])
	addends := strings.Split(leftPart, "+")
	for i := range addends {
		addends[i] = strings.TrimSpace(addends[i])
		if len(addends[i]) == 0 {
			return nil, errors.New("empty addend")
		}
	}
	result := strings.TrimSpace(rightPart)
	if len(result) == 0 {
		return nil, errors.New("empty result")
	}

	// Collect unique letters and track first letters
	letters := []rune{}
	firstLetters := make(map[rune]bool)
	letterSet := make(map[rune]bool)

	// Process addends
	for _, word := range addends {
		// Track first letter of each addend
		firstChar := rune(word[0])
		firstLetters[firstChar] = true

		// Collect all letters in addend
		for _, c := range word {
			if !letterSet[c] {
				letters = append(letters, c)
				letterSet[c] = true
			}
		}
	}

	// Process result word
	firstChar := rune(result[0])
	firstLetters[firstChar] = true
	for _, c := range result {
		if !letterSet[c] {
			letters = append(letters, c)
			letterSet[c] = true
		}
	}

	// Check if there are more than 10 unique letters
	if len(letters) > 10 {
		return nil, errors.New("too many unique letters")
	}

	// Create a map from letter to index for quick lookup
	letterIndex := make(map[rune]int)
	for i, c := range letters {
		letterIndex[c] = i
	}

	// Prepare for backtracking
	solution := make(map[string]int)
	solutionDigits := make([]int, len(letters))
	used := make([]bool, 10)
	found := false

	var backtrack func(int)
	backtrack = func(pos int) {
		if found {
			return
		}
		if pos == len(letters) {
			// Calculate sum of addends
			sum := 0
			for _, word := range addends {
				num := 0
				for _, c := range word {
					num = num*10 + solutionDigits[letterIndex[c]]
				}
				sum += num
			}

			// Calculate result value
			resultNum := 0
			for _, c := range result {
				resultNum = resultNum*10 + solutionDigits[letterIndex[c]]
			}

			if sum == resultNum {
				// Build the solution map
				for i, c := range letters {
					solution[string(c)] = solutionDigits[i]
				}
				found = true
			}
			return
		}

		currentLetter := letters[pos]
		for d := 0; d < 10; d++ {
			if used[d] {
				continue
			}
			// Skip zero for first letters
			if d == 0 && firstLetters[currentLetter] {
				continue
			}
			used[d] = true
			solutionDigits[pos] = d
			backtrack(pos + 1)
			if found {
				return
			}
			used[d] = false
		}
	}

	backtrack(0)

	if found {
		return solution, nil
	}
	return nil, errors.New("no solution")
}
